src/main/java/org/apache/sling/resourcemerger/impl/MergedResourcePickerWhiteboard.java

```diff
@@ -24,102 +24,96 @@
+    }
 
-                serviceRegistrations.put(key, reg);
-            }
-            return pickerObj;
-        }
-        return null;
+    @Reference(policy = ReferencePolicy.DYNAMIC)
+    public void bindMergedResourcePicker2(MergedResourcePicker2 resourcePicker, Map<String, String> properties) {
+        registerMergingResourceProvider(resourcePicker, properties);
     }
 
-    @Override
-    public void modifiedService(final ServiceReference reference, final Object service) {
-        removedService(reference, service);
-        addingService(reference);
+    public void unbindMergedResourcePicker2(Map<String, String> properties) {
+        unregisterMergingResourceProvider(properties);
     }
 
-    @Override
-    public void removedService(final ServiceReference reference, final Object service) {
-        final Long key = (Long) reference.getProperty(Constants.SERVICE_ID);
-        final ServiceRegistration reg = serviceRegistrations.get(key);
-        if ( reg != null ) {
-            reg.unregister();
-            this.bundleContext.ungetService(reference);
+    private void registerMergingResourceProvider(MergedResourcePicker2 resourcePicker, Map<String, String> properties) {
+        final String mergeRoot = properties.getOrDefault(MergedResourcePicker2.MERGE_ROOT, null);
+        if (mergeRoot != null) {
+            boolean readOnly = PropertiesUtil.toBoolean(properties.get(MergedResourcePicker2.READ_ONLY), true);
+            boolean traverseParent = PropertiesUtil.toBoolean(properties.get(MergedResourcePicker2.TRAVERSE_PARENT), false);
+
+            MergingResourceProvider provider = readOnly ?
+                    new MergingResourceProvider(mergeRoot, resourcePicker, true, traverseParent) :
+                    new CRUDMergingResourceProvider(mergeRoot, resourcePicker, traverseParent);
+    
+            final Dictionary<String, Object> props = new Hashtable<>();
+            props.put(ResourceProvider.PROPERTY_NAME, readOnly ? "Merging" : "CRUDMerging");
+            props.put(ResourceProvider.PROPERTY_ROOT, mergeRoot);
+            props.put(ResourceProvider.PROPERTY_MODIFIABLE, !readOnly);
+            props.put(ResourceProvider.PROPERTY_AUTHENTICATE, ResourceProvider.AUTHENTICATE_NO);
+    
+            final ServiceRegistration<ResourceProvider> resourceProvider = bundleContext.registerService(ResourceProvider.class, provider, props);
+            resourceProvidersPerRoot.put(mergeRoot, resourceProvider);
         }
     }
 
+    private void unregisterMergingResourceProvider(Map<String, String> properties) {
+        final String mergeRoot = properties.getOrDefault(MergedResourcePicker2.MERGE_ROOT, null);
+        if (mergeRoot != null) {
+            final ServiceRegistration<ResourceProvider> resourceProvider = resourceProvidersPerRoot.get(mergeRoot);
+            if (resourceProvider != null) {
+                resourceProvider.unregister();
+            }
+        }
```

Comment:
... but clearly, if not, this opens things up to some problems. Let's say you have two services with the same merge root, previously, I think they would have just both be registered (and eventually, unregistered when they go away). Like this, you could have them both register and only one unregister, no?